dashboard/streamlit_app.py

```diff
@@ -1,73 +1,177 @@
+    "entry_id", "species", "temperature", "soil_moisture",
+    "last_watered", "recording_taken", "sunlight", "botanist_name", "cycle"
+]
+
+
+def get_db_connection():
+    """Establishes a connection with the PostgreSQL database."""
+    try:
+        conn = connect(
+            user=environ.get("DATABASE_USERNAME"),
+            password=environ.get("DATABASE_PASSWORD"),
+            host=environ.get("DATABASE_IP"),
+            port=environ.get("DATABASE_PORT"),
+            database=environ.get("DATABASE_NAME"),)
+        print("Database connection established successfully.")
+        return conn
+    except Error as err:
+        print("Error connecting to database: ", err)
+        sys.exit()
+
+
+def get_live_database(conn: connection) -> pd.DataFrame:
+    """Retrieve plant data from the database and return as a DataFrame."""
+
+    query = """
+    SELECT p.plant_entry_id AS entry_id,
+           s.s_name AS species,
+           p.temperature,
+           p.soil_moisture,
+           p.last_watered AS last_watered,
+           p.recording_taken AS recording_taken,
+           sun.s_description AS sunlight,
+           b.b_name AS botanist_name,
+           c.cycle_name AS cycle
+    FROM plant p
+    LEFT JOIN species s ON p.species_id = s.species_id
+    LEFT JOIN sunlight sun ON p.sunlight_id = sun.sunlight_id
+    LEFT JOIN botanist b ON p.botanist_id = b.botanist_id
+    LEFT JOIN cycle c ON p.cycle_id = c.cycle_id;
+    """
+
+    with conn.cursor() as cur:
+        cur.execute(query)
+        data = cur.fetchall()
+
+    data_df = pd.DataFrame(data, columns=CSV_COLUMNS)
+    data_df["last_watered"] = pd.to_datetime(data_df["last_watered"])
+    data_df["recording_taken"] = pd.to_datetime(data_df["recording_taken"])
+
+    return data_df
+
+
+def get_selected_archive() -> pd.DataFrame:
+    """ Creates the dropdown and which archived file 
+    to select and return for display."""
+    with st.sidebar:
+        st.sidebar.title("Dropdown")
+
+        csv_files = [f for f in os.listdir(
+            "archived_data") if f.endswith('.csv')]
+
+        # Dropdown to select a CSV file.
+        selected_csv = st.selectbox(
+            "Select an archived file.", csv_files, on_change=on_toggle_or_archive_change)
+        data = pd.read_csv(os.path.join("archived_data", selected_csv))
+    return data
 
-def get_live_data(db_conn: connection) -> pd.DataFrame:
-    """Function that gets the required data from the database connection 
-    and puts all data into pandas DataFrame form"""
-    cursor = db_conn.cursor()
-    cursor.execute("SELECT * FROM plant")
-    names = [x[0] for x in cursor.description]
-    live_plants_data = pd.DataFrame(cursor.fetchall(), columns=names)
-    return live_plants_data
 
+def on_toggle_or_archive_change() -> None:
+    """
+    Sets the index back to 0 for all of the graphs upon changing
+    the toggle status or which archive is being viewed
+    """
+    st.session_state.start_index = 0
```

Comment:
should this be a return statement and if not why can this not directly be written to where on_toggle_or_archive_change is called?

Reply:
putting the statement directly in the onchange gives a syntax error, and I can return it but I am not sure if returning has more functional change as it is an extra line to do it and I'm not using the return of that function anywhere else. Feel free to let me know what you think?

Reply:
The crowd has spoken -- I think it's fine -- sorry to ask so many questions, my brain was not understanding!